feat(api): (1) Upgrade Upload APIs to Version 2

dvjsharma opened the pull request:
<!-- SPDX-FileCopyrightText: © Fossology contributors

     SPDX-License-Identifier: GPL-2.0-only
-->

<!-- Please refer to CONTRIBUTING.md (https://github.com/fossology/fossology/blob/master/CONTRIBUTING.md)
before creating the pull request to make sure you follow all the standards. -->

## Description

This PR addresses tasks 1, 3, 4, 5, 6, 9, 10, and 11 outlined in issue #2623 

### Changes

**Common for all: Moved `groupName` variable from headers to query parameters by making a change in  RestAuthMiddleware**

**Task 1: /updoads/{id}, GET:**
 - moved `groupName` variable  from headers to query params
 - changed the response keys to camelCase
 - modified the unit test to accommodate changes  

**Task 3: /updoads/{id}, DELETE:**

-  moved `groupName` variable  from headers to query params

**Task 4: /updoads/{id}, PATCH:**

-  moved `groupName` variable  from headers to query params

**Task 5: /updoads/{id}, PUT:**
-  moved `groupName`, `folderId`, `action` variable  from headers to query params
-  modified the unit test to accommodate changes  

**Task 6: /updoads, GET:** 
- moved `groupName`, `page`, `limit` variable  from headers to query params

**Task 9: /uploads/{id}/item/{itemId}/info, GET:**
- changed the response keys to camelCase 

**Task 10: /uploads/{id}/licenses, GET:**
-  moved `groupName`, `page`, `limit` variable  from headers to query params 
-  changed the response keys to camelCase 

**Task 11: /uploads/{id}/perm-groups, GET:**
-  updated the API Url to `/uploads/{id}/groups/permission`
-  changed the response keys to camelCase 

## How to test

- Changes pass all the unit tests
- Can be tested manually by making requests to the mentioned endpoints using swagger or postman

/cc @dushimsam @GMishx @shaheemazmalmmd

dushsam2 commented:
Hello @dvjsharma ,
Would you please squash your commits into one?

dushimsam commented:
Hello @dvjsharma , would you please squash these commits into a one?

dvjsharma commented:
Done. Please take a look, @dushimsam

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -207,8 +207,7 @@ public function getUploads($request, $response, $args)
     }
-
-    $page = $request->getHeaderLine(self::PAGE_PARAM);
+    $page = $query[self::PAGE_PARAM] ?? "";
```
The API should still support V1 requests in deprecated mode until it is completely removed. You can read the version of request with `ApiVersion::getVersion($request) == ApiVersion::V2`.

See https://github.com/fossology/fossology/pull/2617/files#diff-1bae1ab555e908fff79a5b02ff0f05677a8d1f5653259e98531071acfeb7e16bR42 for reference.

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -376,7 +375,8 @@ public function deleteUpload($request, $response, $args)
   {
-    $action = $request->getHeaderLine('action');
+    $queryParams = $request->getQueryParams();
+    $action = $queryParams['action'] ?? ""; 
```
The API should still support V1 requests in deprecated mode until it is completely removed. You can read the version of request with `ApiVersion::getVersion($request) == ApiVersion::V2`.

See https://github.com/fossology/fossology/pull/2617/files#diff-1bae1ab555e908fff79a5b02ff0f05677a8d1f5653259e98531071acfeb7e16bR42 for reference.

GMishx commented on src/www/ui/api/Controllers/UploadController.php:
```diff
@@ -810,6 +811,10 @@ public function getGroupsWithPermissions($request, $response, $args)
     $finalPermGroups = [];
     foreach ($permGroups as $value) {
       unset($value["perm_upload_pk"]);
+      $value["groupPk"] = $value["group_pk"];
+      unset($value["group_pk"]);
+      $value["groupName"] = $value["group_name"];
+      unset($value["group_name"]);
       $finalPermGroups[] = $value;
```
I would instead recommend to create a new model and get the version aware array from it. See https://github.com/fossology/fossology/pull/2617/files#diff-f62f1d2b860c504c87d386cdfa79513c109848593f0afeb60a5dc76e4b98ff42 for reference.

GMishx commented on src/www/ui/api/Models/FileInfo.php:
```diff
@@ -85,11 +85,11 @@ public function getJSON()
-      'package_info' => $this->package_info,
-      'tag_info' => $this->tag_info,
-      'reuse_info' => $this->reuse_info
+      'viewInfo' => $this->view_info,
+      'metaInfo' => $this->meta_info,
+      'packageInfo' => $this->package_info,
+      'tagInfo' => $this->tag_info,
+      'reuseInfo' => $this->reuse_info
```
I would instead recommend to make the model version aware. See https://github.com/fossology/fossology/pull/2617/files#diff-f62f1d2b860c504c87d386cdfa79513c109848593f0afeb60a5dc76e4b98ff42 for reference.

GMishx commented on src/www/ui/api/Models/Upload.php:
```diff
@@ -96,12 +96,12 @@ public function getJSON()
-      "foldername"  => $this->folderName,
+      "folderId"    => $this->folderId,
+      "folderName"  => $this->folderName,
       "id"          => $this->uploadId,
       "description" => $this->description,
-      "uploadname"  => $this->uploadName,
-      "uploaddate"  => $this->uploadDate,
+      "uploadName"  => $this->uploadName,
+      "uploadDate"  => $this->uploadDate,
       "assignee"    => $this->assignee,
       "hash"        => $this->hash->getArray()
```
I would instead recommend to make the model version aware. See https://github.com/fossology/fossology/pull/2617/files#diff-f62f1d2b860c504c87d386cdfa79513c109848593f0afeb60a5dc76e4b98ff42 for reference.

GMishx commented on src/www/ui/api/documentation/openapiv2.yaml:
```diff
@@ -23,7 +23,7 @@ security:
 externalDocs:
   description: Basic guide
-  url: https://github.com/fossology/fossology/wiki/FOSSology-REST-API
+  url: https://github.com/fossolo/gy/fossology/wiki/FOSSology-REST-API
```
Irrelevant change. Please revert.

GMishx commented on src/www/ui_tests/api/Models/UploadTest.php:
```diff
@@ -30,12 +30,12 @@ public function testDataFormat()
-      "foldername"  => 'root',
+      "folderId"    => 2,
+      "folderName"  => 'root',
       "id"          => 3,
       "description" => '',
-      "uploadname"  => 'my.tar.gz',
-      "uploaddate"  => '01-01-2020',
+      "uploadName"  => 'my.tar.gz',
+      "uploadDate"  => '01-01-2020',
       "assignee"    => 3,
       "hash"        => $hash->getArray()
```
Please make the tests version aware. See https://github.com/fossology/fossology/pull/2617/files#diff-97059fc9a26e02adf4caa85aa5714fe79354dd2c7a33b9d5b303e8be10404834 for reference.

GMishx reviewed (changes requested):
Please make the changes version aware as v1 is to be supported in deprecated mode.

dvjsharma commented:
Hello @GMishx , I have made the following changes as requested: 

-  Added support for both API versions, now  V1 requests can also be made until it is completely removed.
-  Created 2 new models namely `GroupPermission.php` and `Permissions.php` as requested and modified existing models to make them version aware and respond accordingly.
-  Modified unit tests to make them version aware.

Please take a look and suggest changes, Thanks!

dvjsharma commented:
Hi @GMishx , I've made changes for the tests to work. Please review, Thanks!

dvjsharma commented:
Hi @GMishx , Friendly reminder to review the latest changes in the PR. Your input is appreciated. Thanks!

GMishx commented on src/www/ui/api/Models/GroupPermission.php:
```diff
@@ -0,0 +1,123 @@
+  public function getArray($version=ApiVersion::V1)
+  {
+    if ($version == ApiVersion::V2) {
+        return [
+            'perm' => $this->getPerm(),
+            'groupPk' => $this->getGroupPk(),
+            'groupName' => $this->getGroupName()
+        ];
```
Please use 2 space indentation.

GMishx commented on src/www/ui/api/Models/Permissions.php:
```diff
@@ -0,0 +1,95 @@
+   */
+  public function getArray($version=ApiVersion::V1)
+  {
+    if ($version == ApiVersion::V2) {
+        return [
+            'publicPerm' => $this->getpublicPerm(),
+            'permGroups' => $this->getpermGroups(),
+        ];
+    }
+    return [
+        'publicPerm' => $this->getpublicPerm(),
+        'permGroups' => $this->getpermGroups(),
+    ];
```
Please use 2 space indentation.

GMishx reviewed (changes requested):
Minor changes. Rest looks good.

dvjsharma commented:
Hi @GMishx , I've done the indentation changes, PTAL 
Thanks!

GMishx reviewed (approved):
Changes looks good.

GMishx reviewed (approved):
Tested, working as expected.
